au4/src/projectscene/view/clipsview/au3/au3wavepainter.cpp

```diff
@@ -616,22 +617,40 @@ void Au3WavePainter::paint(QPainter& painter, const trackedit::ClipKey& clipKey,
     //     return;
     // }
     // LOGD() << "trackId: " << clipKey.trackId << ", clip: " << clipKey.index;
+    WaveTrack* origWaveTrack = DomAccessor::findWaveTrack(projectRef(), TrackId(clipKey.trackId));
 
     //! Pending tracks are same as project tracks, but with new tracks when recording, so we need draw them
-    Track* track = &PendingTracks::Get(projectRef())
+    Track* pendingTrack = &PendingTracks::Get(projectRef())
                    .SubstitutePendingChangedTrack(*DomAccessor::findWaveTrack(projectRef(), TrackId(clipKey.trackId)));
 
-    WaveTrack* waveTrack = dynamic_cast<WaveTrack*>(track);
-    IF_ASSERT_FAILED(waveTrack) {
+    WaveTrack* pendingWaveTrack = dynamic_cast<WaveTrack*>(pendingTrack);
+    IF_ASSERT_FAILED(pendingWaveTrack) {
         return;
     }
 
-    std::shared_ptr<WaveClip> clip = DomAccessor::findWaveClip(waveTrack, clipKey.clipId);
+    // Iterate through both tracks simultaneously to find matching clip
+    // in the original track and get its counterpart in the pending track.
+    std::optional<uint64_t> clipIdToUpdate;
+    auto origIter = origWaveTrack->Intervals().begin();
+    auto pendingIter = pendingWaveTrack->Intervals().begin();
+
+    for (; origIter != origWaveTrack->Intervals().end() && pendingIter != pendingWaveTrack->Intervals().end(); ++origIter, ++pendingIter) {
```

Comment:
I think we need to put this logic somewhere so that it can be reused (it's already been used twice in this PR)
Something like this
```
ClipId DomAccessor::findMatchedClip(const WaveTrack* track, const WaveTrack* originTrack, const ClipId& originClipId) const 
{
...
}
```

I would do the implementation like this
```
size_t idx = DomAccessor::findClipIndexById(const WaveTrack* track,   const ClipId& clipId);  
if (idx == muse::nidx) {
   // not found 
   return ClipId();
}

ClipId id = DomAccessor::findClipIdByIndex(const WaveTrack* track,  size_t clipIndex); 

```

findClipIndexById, findClipIdByIndex - need add implementation 

but this is just advice, I don't insist, it seems clearer to me, and the findClipIndexById and findClipIdByIndex methods may still be useful to us